### PR TITLE
Publish unified points

### PR DESCRIPTION
**Pull request by lpk1950**
- added a publisher for unified pointcloud
 - added a param 'point_cloud' for requesting to publish the unified scan as unified pointcloud

### CODE REVIEW

File: cob_scan_unifier/src/scan_unifier_node.cpp [hunk 1]
```diff
@@ -28,6 +28,7 @@ ScanUnifierNode::ScanUnifierNode()
 
   // Publisher
   topicPub_LaserUnified_ = nh_.advertise<sensor_msgs::LaserScan>("scan_unified", 1);
+  topicPub_PointcloudUnified_ = nh_.advertise<sensor_msgs::PointCloud2>("/pointcloud_unified", 1);
```
Comment (jabbenseth) [inline cob_scan_unifier/src/scan_unifier_node.cpp]: use the same namespace as for scan_unified
```suggestion
  topicPub_PointcloudUnified_ = nh_.advertise<sensor_msgs::PointCloud2>("pointcloud_unified", 1);
```

File: cob_scan_unifier/src/scan_unifier_node.cpp [hunk 2]
```diff
@@ -113,11 +114,22 @@ ScanUnifierNode::~ScanUnifierNode()
 void ScanUnifierNode::getParams()
 {
   std::vector<std::string> topicList;
+  bool point_cloud;
 
   if (pnh_.getParam("input_scans", topicList))
   {
     config_.input_scan_topics = topicList;
     config_.number_input_scans = config_.input_scan_topics.size();
+    if(!pnh_.hasParam("point_cloud"))
```
Comment (jabbenseth) [inline cob_scan_unifier/src/scan_unifier_node.cpp]: maybe for better semantic naming call your parameter `publish_point_cloud` or something alike

also why is the parameter checked in the 
```
if (pnh_.getParam("input_scans", topicList))
{
```
section and not in the "root" of `getParams()`

File: cob_scan_unifier/src/scan_unifier_node.cpp [hunk 3]
```diff
@@ -140,13 +152,18 @@ void ScanUnifierNode::messageFilterCallback(const sensor_msgs::LaserScan::ConstP
   current_scans.push_back(scan2);
 
   sensor_msgs::LaserScan unified_scan = sensor_msgs::LaserScan();
+  sensor_msgs::PointCloud2 unified_point = sensor_msgs::PointCloud2();
   if (!unifyLaserScans(current_scans, unified_scan))
   {
     return;
   }
-
+  projector_.transformLaserScanToPointCloud(frame_, unified_scan, unified_point, listener_);
```
Comment (jabbenseth) [inline cob_scan_unifier/src/scan_unifier_node.cpp]: this is only needed if config_.pub_point_cloud is true?

File: cob_scan_unifier/src/scan_unifier_node.cpp [hunk 4]
```diff
@@ -179,14 +201,19 @@ void ScanUnifierNode::messageFilterCallback(const sensor_msgs::LaserScan::ConstP
...
   ROS_DEBUG("Publishing unified scan.");
   topicPub_LaserUnified_.publish(unified_scan);
+  if(config_.pub_point_cloud)
+  {
+  topicPub_PointcloudUnified_.publish(unified_point);
+  }
```
Comment (jabbenseth) [inline cob_scan_unifier/src/scan_unifier_node.cpp:210]: there is a lot of duplicated code here. is it possible to move that to "unifyLaserScans" or create a "publish" that handles publishing scan and point cloud

**Review by jabbenseth — CHANGES_REQUESTED** [review]

File: cob_scan_unifier/include/cob_scan_unifier/scan_unifier_node.h [hunk 5]
```diff
@@ -58,6 +58,7 @@ class ScanUnifierNode
     struct config_struct{
       int number_input_scans;
       std::vector<std::string> input_scan_topics;
+      bool pub_point_cloud;
```
Comment (fmessmer) [inline cob_scan_unifier/include/cob_scan_unifier/scan_unifier_node.h]: as suggested by @jabbenseth - rename to `publish_pointcloud`

File: cob_scan_unifier/src/scan_unifier_node.cpp [hunk 6]
```diff
@@ -113,11 +114,22 @@ ScanUnifierNode::~ScanUnifierNode()
 void ScanUnifierNode::getParams()
 {
   std::vector<std::string> topicList;
+  bool point_cloud;
```
Comment (fmessmer) [inline cob_scan_unifier/src/scan_unifier_node.cpp]: get rid of this local helper variable
you can directly write `config_.publish_pointcloud` in both if and else case respectively

File: cob_scan_unifier/src/scan_unifier_node.cpp [hunk 7]
```diff
@@ -113,11 +114,22 @@ ScanUnifierNode::~ScanUnifierNode()
...
     config_.input_scan_topics = topicList;
     config_.number_input_scans = config_.input_scan_topics.size();
+    if(!pnh_.hasParam("point_cloud"))
+    {
+    ROS_WARN("No parameter set for publishing point cloud. Using default value [False].");
+    point_cloud = false;
+    }
+    else
+    {
+    pnh_.getParam("point_cloud", point_cloud);
```
Comment (fmessmer) [inline cob_scan_unifier/src/scan_unifier_node.cpp]: indentation is wrong

File: cob_scan_unifier/src/scan_unifier_node.cpp [hunk 8]
```diff
@@ -140,13 +152,18 @@ void ScanUnifierNode::messageFilterCallback(const sensor_msgs::LaserScan::ConstP
...
   {
     return;
   }
-
+  projector_.transformLaserScanToPointCloud(frame_, unified_scan, unified_point, listener_);
   ROS_DEBUG("Publishing unified scan.");
   topicPub_LaserUnified_.publish(unified_scan);
+  if(config_.pub_point_cloud)
+  {
+  topicPub_PointcloudUnified_.publish(unified_point);
```
Comment (fmessmer) [inline cob_scan_unifier/src/scan_unifier_node.cpp]: indentation is wrong

File: cob_scan_unifier/src/scan_unifier_node.cpp [hunk 9]
```diff
@@ -158,14 +175,19 @@ void ScanUnifierNode::messageFilterCallback(const sensor_msgs::LaserScan::ConstP
...
```
Comment (fmessmer) [inline cob_scan_unifier/src/scan_unifier_node.cpp]: indentation is wrong

File: cob_scan_unifier/src/scan_unifier_node.cpp [hunk 10]
```diff
@@ -179,14 +201,19 @@ void ScanUnifierNode::messageFilterCallback(const sensor_msgs::LaserScan::ConstP
...
```
Comment (fmessmer) [inline cob_scan_unifier/src/scan_unifier_node.cpp]: indentation is wrong

**Review by fmessmer — CHANGES_REQUESTED** [review]
in addition to @jabbenseth review

File: cob_scan_unifier/include/cob_scan_unifier/scan_unifier_node.h [hunk 11]
```diff
@@ -135,6 +135,7 @@ class ScanUnifierNode
      * output:
      * @param: a laser scan message containing unified information from all scanners
      */
+    void Publish(sensor_msgs::LaserScan& unified_scan);
```
Comment (jabbenseth) [inline cob_scan_unifier/include/cob_scan_unifier/scan_unifier_node.h]: methods should be lower case 
(publish)

**Review by jabbenseth — CHANGES_REQUESTED** [review]

**Review by jabbenseth — DISMISSED** [review]

**Review by fmessmer — APPROVED** [review]

**Pull request merged**